### ISSUE TITLE
Make the GUI better and more user friendly

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#247]
Many (probably the majority) of Abel transform users are not Python users. So, it's likely that many more people would use PyAbel if our GUI was more capable and easier to use. I think that there are two parts to this:

1) Making the GUI itself better.
2) Improving the documentation (in the readme) so that non-python users can figure out how to install Python, install PyAbel, and run the GUI.

(One could argue that we should really get out act together and figure out how to make a packaged executable, but with out limited resources I'd say that we stick with a simple python script as out GUI.)

Here is what our current GUI (which lives at examples/example_GUI.py) looks like:

![image](https://user-images.githubusercontent.com/1107796/54546483-594b1f80-4969-11e9-9059-42a59903ce49.png)

I think that it's a good start, but could certainly use some improvement. 

Can someone post some screenshots of the old basex.exe program so that we can make sure that we have, at least, some answer to the major functionality of that program? 

I think that we should start by describing the major features and functionality that we'd like to see in our GUI, and then we can think about what it should actually look like.

**Comment by MikhailRyazanov** [comment]
In my case (GNU/Linux, 227 dpi screen) it looks much worse:
![20190318-165430](https://user-images.githubusercontent.com/12854133/54568991-efe60380-499e-11e9-9f3b-2e92398edeaf.png)
Besides that, the current version crashes and/or complains about deprecated functions (in different places, depending on the Python version).

I'll post BASEX screenshots later, but everybody please also post screenshots of what you are actually using in your work. This especially pertains to those who study flames/plasmas, since you do not need the “speed” and “anisotropy” distributions, but maybe need something else.

**Comment by MikhailRyazanov** [comment]
Here are the BASEX screenshots:

“Initial image”:
![1](https://user-images.githubusercontent.com/12854133/54571196-3ccdd800-49a7-11e9-81a1-52fa821c9c50.png)
The image is selected and loaded by clicking the folder icon. The crossed bars invokes “Image Centering”. The gear icon invokes “Parameters of Reconstruction”. The “Rescale Intensity” slider changes the intensity scale.

“Image Centering”:
![2](https://user-images.githubusercontent.com/12854133/54571236-68e95900-49a7-11e9-868e-0b40a756bbbf.png)
Has automatic methods and manual controls.

“Parameters of Reconstruction”:
![3](https://user-images.githubusercontent.com/12854133/54571277-96ce9d80-49a7-11e9-8a63-e4dbe94df644.png)
After clicking “OK” start all the calculations.

“Reconstructed Image”:
![4](https://user-images.githubusercontent.com/12854133/54571315-ced5e080-49a7-11e9-925c-9c9f51ecc284.png)

“Speed Distribution”:
![5](https://user-images.githubusercontent.com/12854133/54571351-f2009000-49a7-11e9-8cba-788795677da2.png)

The anisotropy distribution is not displayed. All the results are saved automatically (see “Parameters of Reconstruction”).

**Comment by MikhailRyazanov** [comment]
Here are the BASEX screenshots:

“Initial image”:
![1](https://user-images.githubusercontent.com/12854133/54571196-3ccdd800-49a7-11e9-81a1-52fa821c9c50.png)
The image is selected and loaded by clicking the folder icon. The crossed bars invokes “Image Centering”. The gear icon invokes “Parameters of Reconstruction”. The “Rescale Intensity” slider changes the intensity scale.

“Image Centering”:
![2](https://user-images.githubusercontent.com/12854133/54571236-68e95900-49a7-11e9-868e-0b40a756bbbf.png)
Has automatic methods and manual controls.

“Parameters of Reconstruction”:
![3](https://user-images.githubusercontent.com/12854133/54571277-96ce9d80-49a7-11e9-8a63-e4dbe94df644.png)
After clicking “OK” start all the calculations.

“Reconstructed Image”:
![4](https://user-images.githubusercontent.com/12854133/54571315-ced5e080-49a7-11e9-925c-9c9f51ecc284.png)

“Speed Distribution”:
![5](https://user-images.githubusercontent.com/12854133/54571351-f2009000-49a7-11e9-8cba-788795677da2.png)

The anisotropy distribution is not displayed. All the results are saved automatically (see “Parameters of Reconstruction”).

**Comment by stggh** [comment]
@DanHickstein & @MikhailRyazanov, you guys certainly press the accelerator ;-)

(1) basex.exe  (run on Linux using WINE):
![Screenshot-20190319104851-866x661](https://user-images.githubusercontent.com/10932229/54571296-3aff2580-4a36-11e9-97a3-f2c27bbd47b0.png)

 ![Screenshot-20190319104918-866x661](https://user-images.githubusercontent.com/10932229/54571003-eb6c2a00-4a34-11e9-8efb-e49997c2af39.png)

(2) PyAbel GUI
This was written for fun. It lacks a `save` button.

It no longer runs on my Linux system, probably due to matplotlib:
```
 [etalon:.+as-fix/examples]$ python example_GUI.py 
Traceback (most recent call last):
  File "example_GUI.py", line 22, in <module>
    from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg,\
ImportError: cannot import name 'NavigationToolbar2TkAgg' from 'matplotlib.backends.backend_tkagg' (/home/stg121/Software/Python/anaconda3/lib/python3.7/site-packages/matplotlib/backends/backend_tkagg.py)
```

(3) Features
    (a) image input and display
    (b) centre
    (c) circularize
    (d) Abel transform
    (e) radial profile
    (f) angular profile
    (g) save - at any stage

@MikhailRyazanov - my post held up by crook mouse.

**Comment by MikhailRyazanov** [comment]
@stggh, did I delete your comment? If yes, I apologize — my network connection was disappearing, and I thought that I've posted my comment two times...

Regarding `NavigationToolbar2TkAgg`, see my updated [comment above](https://github.com/PyAbel/PyAbel/issues/247#issuecomment-474135320).

What interface do you actually use in your work?

**Comment by stggh** [comment]
@MikhailRyazanov thanks, `2Tk` does the trick:
![Screenshot-20190319114747-1053x867](https://user-images.githubusercontent.com/10932229/54572684-e8753780-4a3c-11e9-9032-83747763ee98.png)

I have to admit the GUI is fun to play with ;-)

> What interface 

Linux only!
```[etalon:.+as-fix/examples]$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic
```
+ Anaconda python3


> I delete your comment

No, my own mouse froze. It took a while to track down another, and in the meantime you posted the `basex.exe` shots (better than mine BTW).

**Comment by MikhailRyazanov** [comment]
>> What interface
>
> Linux only!

Do you mean just CLI / IPython? :–)

Can anybody explain why I see [this](https://github.com/PyAbel/PyAbel/issues/247#issuecomment-474135320) disordered interface and only one figure (image or plot) in both Debian and Win10? Are we using the same https://github.com/PyAbel/PyAbel/blob/master/examples/example_simple_GUI.py?

**Comment by stggh** [comment]
Try `example_GUI.py`, the *simple* example has a simple interface.

I am using Anaconda python(3) CLI:   `python example_GUI.py`

**Comment by MikhailRyazanov** [comment]
> Try `example_GUI.py`

Oops, my bad! I didn't see it there...

But do we actually need that “simple” GUI example? I think, it's probably better to remove it and move `examples/example_GUI.py` to `abel/GUI.py`, such that it can be executed as `python -m abel.GUI`.

**Comment by stggh** [comment]
> move  `examples/example_GUI.py` to `abel/GUI.py`

Perhaps, when it has had its makeover

**Comment by MikhailRyazanov** [comment]
> Perhaps, when it has had its makeover

Now I'm not sure that the GUI should be an integral part of the module:

The program must have a capability to load and save basis sets, preferably from/to a user-specified directory. And remember this directory and other preferences. This would require at least one “standard” user-writable directory. Does Python have a portable way to achieve this for modules?

If not, it might be better to have the GUI as a separate project. Then it at least could use its own directory by default (a “portable”/“standalone” configuration) or be packaged for particular systems (use `/usr/bin/`, `/var/cache/` and `~/.config/` in FHS systems and whatever their analogs are in Windows and macOS).

**Comment by DanHickstein** [comment]
I really like the idea that users could invoke the GUI from any folder with the `python -m abel.GUI` call. Potentially, could we add a `main()` function to `abel/__init__.py` that would call the GUI, allowing users to just type `python -m abel`?

Finding a place to store the preferences and basis sets is a general problem with applications, which (I believe) can not generally write to the directory where they are located. 

I think that we could store settings in `~/.config/pyabel`, and somewhere similar on Windows systems. See the second answer here:
https://softwareengineering.stackexchange.com/questions/3956/best-way-to-save-application-settings

If users go crazy with basis sets, there is the potential to waste a lot of harddrive space. We may want to make this obvious to the user by making a visible directory: `~/pyabel_basis` or something. 

At this point, I'd like to see a single project for the main PyAbel and the GUI, since it should make it simpler to ensure compatibility, and put all of the conversations in the same place.

**Comment by DanHickstein** [comment]
@DhrubajyotiDas, do you have any requests for pre/post-processing tools or other features that would be valuable to the flame community? Otherwise us photoelectron freaks are just going to fill the GUI with angular integrations, anisotropy measurements, and other clutter that doesn't apply to flames :)

**Comment by DhrubajyotiDas** [comment]
@DanHickstein Would certainly like some tool that can fold an image along the axis of symmetry / specify or draw a central axis around which to do the transform - basically visually determine the proper axis of symmetry if the data is offset laterally within the image.

I have other ideas in mind but those can be a continuous improvement type thing and not gate releasing the GUI for the paper :)

**Comment by MikhailRyazanov** [comment]
@DhrubajyotiDas, centering is needed for the VMI data as well, but it both directions, so this means that there should be an option to skip vertical centering.

Mentioning other ideas is also important, since even if we are not going to implement them immediately, keeping them in mind will help to design the overall structure properly.

**Comment by MikhailRyazanov** [comment]
@DanHickstein,
> `python -m abel`

Yeah, if there is no other use for this, then starting the GUI this way would be the best. If we can use the “standard” config/data paths, then there should be also no problem with the writable files. Is there some Python module to get these paths portably, or we'll have to hard-code them for each OS?

> We may want to make this obvious to the user by making a visible directory: `~/pyabel_basis` or something.

This is, I think, untidy (and probably not recommended). We can instead give a method to clear the cache or, better, display its content and let the user to keep/erase individual files.

**Comment by MikhailRyazanov** [comment]
> Is there some Python module to get these paths portably

[appdirs](https://pypi.org/project/appdirs/) should probably work. It is even packaged in Debian ([Python 2](https://packages.debian.org/stretch/python-appdirs), [Python 3](https://packages.debian.org/stretch/python3-appdirs)), so can be considered standard... ;–)

**Comment by DanHickstein** [comment]
> basically visually determine the proper axis of symmetry if the data is offset laterally within the image.

@DhrubajyotiDas, what exactly do you have in mind for how to visually determine the axis of symmetry? Are you thinking something like the "difference between original and reflected image" plots shown in the Basex.exe examples above? I agree that these would be very helpful. 

@MikhailRyazanov, appdirs seems to answer our questions about where to store settings data. We can either use that module, or just take its recommendations for where to save things on various systems.

Regarding where to store the basis sets: this is an issue for both the GUI and the command-line-interface usage of PyAbel. While we could find a solution that works only for the GUI, I think that this might be an opportunity to fix it globally. Currently, we are saving the basis sets in the current working directory, which litters the users computer with basis sets and most users will likely end up re-generating basis sets needlessly. Saving them in a single user-accessible directory provides a straightforward way for users of both the GUI and the CLI to manage their basis sets.

One more thought about saving basis sets: what happens if a user upgrades their PyAbel after we have implemented a change that breaks backwards compatibility of the basis set files? Should the basis sets be saved with PyAbel version information? Should PyAbel check the "modified" datetime of the basis set file and ensure that this time came after the modified datetime of the `<method>.py` file?

**Comment by DhrubajyotiDas** [comment]
@DanHickstein Here's a quick mock-up of what I was thinking.

There will be a GUI with a window on the left side that shows a 1D trace (or multiple traces) with a user defined axis of symmetry that can be moved left or right, and a live updating delta window on the right that shows the (signed) difference between the two sides along the length of axis of symmetry. The user can either accept the default axis, which would be at the geometric center of the window on the left, or shift it towards one side or the other as needed. 

![image](https://user-images.githubusercontent.com/6530840/55296427-1ad44c80-53ce-11e9-9058-19a7dbfa1364.png)

**Comment by DanHickstein** [comment]
Oh, that's interesting. Not the same as what basex.exe is doing, since this is giving you the information for every point in the vertical, but only for the currently selected center. I suppose we could also show the basex.exe plot, which shows the integrated difference for each option for the center position.

What is the advantage of showing this plot instead of a single number? Does it let you diagnose issues with the image, such as a slight rotation that need to be addressed?

**Comment by DhrubajyotiDas** [comment]
> Does it let you diagnose issues with the image, such as a slight rotation that need to be addressed?

Exactly. 

Usually it's a rotation issue that is easier to fix by dinking with my experimental setup (crooked cameras and/or stages) and collect new data instead of trying to salvage the old image. Sometimes it is a translation issue, which I can fix without needing a new image by moving the center axis laterally (or equivalently, chopping off columns of pixels from either side as needed).

**Comment by MikhailRyazanov** [comment]
By the way, we can also add a possibility to rotate the image. At least by 90° (some people might have the symmetry axis being horizontal), which will be very easy. But arbitrary rotation might be also useful. I was recently asked by other people in our group to help them with implementing BASEX for their system. They study nanoparticles and observe something presumably cylindrically symmetric but with the axis defined by the particle orientation, which can be arbitrary (so they actually have a tool to rotate their images).

**Comment by DanHickstein** [comment]
Including image rotation seems fairly easy. We can basically just use [scipy.ndimage.rotate](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.rotate.html), and we need some sort of check box to select which `reshape=` behavior we want.

**Comment by MikhailRyazanov** [comment]
By the way, recently I was using PyAbel's image circularization tools for an unrelated purpose, to rectify Fabry–Pérot fringes and convert them to a spectrum, and made a small GUI for this purpose (some [code](https://gist.github.com/MikhailRyazanov/3a37b860481a8ca5d48625311ba1ecfb) to demonstrate the idea; use arrow keys to move the center, with Ctrl to move faster).

This is how it looks
<table>
 <tr><td> original image </td><td> circularized in polar </td></tr>
 <tr><td> circularized image </td><td> radial profile </td></tr>
</table>

when poorly centered:

![bad](https://user-images.githubusercontent.com/12854133/58224343-9d94cc00-7cda-11e9-8469-85bc5d3c1368.png)

and when finally well centered and circularized:

![good](https://user-images.githubusercontent.com/12854133/58224360-b1403280-7cda-11e9-8a4a-f5c6651c2585.png)

Something like this might be useful for analyzing VMI data (when I was doing my VMI, I did similar things using my own rectification code, but non-interactively).

**Comment by stggh** [comment]
I am exhausting all my "neat!"s with your work @MikhailRyazanov 

Here is O<sub>2</sub><sup>-</sup> unoptimized, but it illustrates the radial structure, and residual image distortion. It is possible (with +10 stepping) to observe the transition peak heights change. 

![image](https://user-images.githubusercontent.com/10932229/58225814-412bbe00-7d67-11e9-8095-eeeaa3e959f7.png)

Definitely a worthy inclusion in GUIv2.

**Comment by MikhailRyazanov** [comment]
I guess that for images of reasonable sizes our fastest methods (like `onion_peeling`) can inverse-transform all quadrants faster than the circularization itself takes (`matplotlib` plotting is also not very fast), so we can actually plot the reconstructed polar representation and radial profile(s), which will be sharper and thus easier to use.

**Comment by stggh** [comment]
Good idea. It works well. 

![image](https://user-images.githubusercontent.com/10932229/58302547-068d5880-7e2f-11e9-9143-52df061b1c4d.png)

`ax[1][1]` would be better with static axis limits, for a give zoom, to allow examination of a single peak, while the origin centre is shifted. 

`scipy.find_peaks()` could be employed to provide quantitative peak widths.

**Comment by MikhailRyazanov** [comment]
Yes, the two right panels must be synchronously zoomable. And the grid should be user-definable, with the help of automatic peak finding. Plotting radial profiles along several user-defined angles is useful too (this allows to see how well they are aligned more precisely than the polar image alone).
